official/vision/beta/projects/yolo/optimization/sgd_torch.py

```diff
@@ -0,0 +1,339 @@
+    """
+
+    # pylint: disable=protected-access
+    # Get the distributed variable if it exists.
+    if hasattr(var, "_distributed_container"):
+      var = var._distributed_container()
+    if var._in_graph_mode:
+      return var._shared_name
+    return var._unique_id
+
+
+class SGDTorch(tf.keras.optimizers.Optimizer):
+  """Optimizer that simulates the SGD module used in pytorch. 
+  
+  
+  For details on the differences between the original SGD implemention and the 
+  one in pytorch: https://pytorch.org/docs/stable/generated/torch.optim.SGD.html.
+  This optimizer also allow for the usage of a momentum warmup along side a 
+  learning rate warm up, though using this is not required. 
+
+  Example of usage for training:
+  ```python
+  opt = SGDTorch(learning_rate, weight_decay = 0.0001)
+  l2_regularization = None
+
+  # iterate all model.trainable_variables and split the variables by key 
+  # into the weights, biases, and others.
+  optimizer.search_and_set_variable_groups(model.trainable_variables)
+
+  # if the learning rate schedule on the biases are different. if lr is not set 
+  # the default schedule used for weights will be used on the biases. 
+  opt.set_bias_lr(<lr schedule>)
+
+  # if the learning rate schedule on the others are different. if lr is not set 
+  # the default schedule used for weights will be used on the biases. 
+  opt.set_other_lr(<lr schedule>)
+  ```
+  """
+
+  _HAS_AGGREGATE_GRAD = True
+
+  def __init__(self,
+               weight_decay=0.0,
+               learning_rate=0.01,
+               momentum=0.0,
+               momentum_start=0.0,
+               warmup_steps=1000,
+               nesterov=False,
+               sim_torch=False,
+               name="SGD",
+               weight_keys=["kernel", "weight"],
+               bias_keys=["bias", "beta"],
+               **kwargs):
+    super(SGDTorch, self).__init__(name, **kwargs)
+
+    # Create Hyper Params for each group of the LR
+    self._set_hyper("learning_rate", kwargs.get("lr", learning_rate))
+    self._set_hyper("bias_learning_rate", kwargs.get("lr", learning_rate))
+    self._set_hyper("other_learning_rate", kwargs.get("lr", learning_rate))
+
+    # SGD decay param
+    self._set_hyper("decay", self._initial_decay)
+
+    # Weight decay param
+    self._weight_decay = weight_decay != 0.0
+    self._set_hyper("weight_decay", weight_decay)
+
+    # Enable Momentum
+    self._momentum = False
+    if isinstance(momentum, ops.Tensor) or callable(momentum) or momentum > 0:
+      self._momentum = True
+    if isinstance(momentum, (int, float)) and (momentum < 0 or momentum > 1):
+      raise ValueError("`momentum` must be between [0, 1].")
+    self._set_hyper("momentum", momentum)
+    self._set_hyper("momentum_start", momentum_start)
+    self._set_hyper("warmup_steps", tf.cast(warmup_steps, tf.int32))
+
+    # Enable Nesterov Momentum
+    self.nesterov = nesterov
+
+    # Simulate Pytorch Optimizer
+    self.sim_torch = sim_torch
```

Comment:
Why do we have this? this is a pytorch optimizer.

Reply:
i was doing comparison testing between running the PyTorch equation and running SGD with weight decay instead of regularization. If you want I can remove that

Reply:
done